taichi/analysis/gather_immutable_local_vars.cpp

```diff
@@ -0,0 +1,92 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+
+namespace taichi::lang {
+
+// The GatherImmutableLocalVars pass gathers all immutable local vars as input
+// to the EliminateImmutableLocalVars pass. An immutable local var is an alloca
+// which is stored only once (in the same block) and only loaded after that
```

Comment:
Speaking of this, since the most amount of time spent was on `store_to_load_forwarding` in CFG, we can make a smaller & simpler specialized `store_to_load_forwarding` that only processes alloca load stores in the same block

Reply:
Yep I considered that as well and may try that in the future. The current one is chosen first because its cost-benefit is surprisingly high. Meanwhile, it's at least a global optimization so it sounds not that hacky :-)